retry/retrier.go

```diff
@@ -0,0 +1,209 @@
+func Exponential(base, adjustment time.Duration) (strategy, strategyType) {
+	if base < 1*time.Second {
+		panic("exponential retry strategies must have a base of at least 1 second")
+	}
+
+	return func(r *Retrier) time.Duration {
+		baseSeconds := int(base / time.Second)
+		exponentSeconds := math.Pow(float64(baseSeconds), float64(r.attemptCount))
+		exponent := time.Duration(exponentSeconds) * time.Second
+
+		return adjustment + exponent + r.calculateJitter()
+	}, exponentialStrategy
+}
+
+type retrierOpt func(*Retrier)
+
+// WithMaxAttempts sets the maximum number of retries that a retrier will attempt
+func WithMaxAttempts(maxAttempts int) retrierOpt {
+	return func(c *Retrier) {
+		c.maxAttempts = maxAttempts
+	}
+}
+
+// WithStrategy sets the retry strategy that the retrier will use to determine how long to wait between retries
+func WithStrategy(strategy strategy, strategyType strategyType) retrierOpt {
+	return func(c *Retrier) {
+		c.strategyType = strategyType
+		c.intervalCalculator = strategy
+	}
+}
+
+// WithJitter enables jitter on the retrier, which will cause all of the retries to wait a random amount of time < 1 second
+// The idea here is to avoid thundering herds - retries that are in parallel will happen at slightly different times when
+// jitter is enabled, whereas if jitter is disabled, all the retries might happen at the same time, causing further load
+// on the system that we're tryung to do something with
+func WithJitter() retrierOpt {
+	return func(c *Retrier) {
+		c.jitter = true
+	}
+}
+
+// TryForever causes the retrier to to never give up retrying, until either the operation succeeds, or the operation
+// calls retrier.Break()
+func TryForever() retrierOpt {
+	return func(c *Retrier) {
+		c.forever = true
+	}
+}
+
+// NewRetrier creates a new instance of the Retrier struct. Pass in retrierOpt functions to customise the behaviour of
+// the retrier
+func NewRetrier(opts ...retrierOpt) *Retrier {
+	r := &Retrier{}
+
+	for _, o := range opts {
+		o(r)
+	}
+
+	// We use panics here rather than returning an error because all of these are logical issues caused by the programmer,
+	// they should never occur in normal running, and can't be logically recovered from
+	if r.maxAttempts == 0 && !r.forever {
+		panic("retriers must either run forever, or have a maximum attempt count")
+	}
+
+	if r.maxAttempts < 0 {
+		panic("retriers must have a positive max attempt count")
+	}
+
+	oldJitter := r.jitter
+	r.jitter = false // Temporarily turn off jitter while we check if the interval is 0
+	if r.forever && r.strategyType == constantStrategy && r.NextInterval() == 0 {
+		panic("retriers using the constant strategy that run forever must have an interval")
+	}
+	r.jitter = oldJitter // and now set it back to what it was previously
+
+	return r
+}
+
+func (r *Retrier) calculateJitter() time.Duration {
+	if r.jitter {
+		return time.Duration(random.Float32()) * jitterInterval
```

Comment:
This can go down as well as up. Does everything correctly handle `NextInterval` returning a negative number?

Reply:
`NextInterval` is only ever used as an arg to time.Sleep, which just [returns immediately](https://github.com/golang/go/blob/master/src/runtime/time.go#L180=) if it's passed a negative interval, so if things go negative we should be all good.

can you explain what you mean by going down as well as up though? `random.Float64` [returns 0 or a positive float](https://pkg.go.dev/math/rand#Float64), though i actually do think that it might be good to have negative jitter be a thing

Reply:
Oh, I just assumed `random.Float64` would return any random float64. That's what you get for not reading the docs. Thanks for letting me know!